# HG changeset patch
# Parent  41a800a10f56ae750a3a327463422f3c6cfd2544

diff -r 41a800a10f56 src/share/vm/prims/jvm.cpp
--- a/src/share/vm/prims/jvm.cpp	Thu Jun 18 15:13:34 2015 -0700
+++ b/src/share/vm/prims/jvm.cpp	Thu Jun 18 15:14:09 2015 -0700
@@ -4235,6 +4235,10 @@
   return DTraceJSDT::is_supported();
 JVM_END
 
+JVM_ENTRY_NO_ENV(void, JVM_SetVmMemoryPressure(jint pressure))
+  // Do nothing, this function is only to make management.dll happy
+JVM_END
+
 // Returns an array of all live Thread objects (VM internal JavaThreads,
 // jvmti agent threads, and JNI attaching threads  are skipped)
 // See CR 6404306 regarding JNI attaching threads
